### PR TITLE
Odin configure api new

### PR DESCRIPTION
**Pull request by rahulrayal**
# Odin version(x.y.z)

### CODE REVIEW

File: app/structure.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	ConfigFile:   "config",
+	EnvVarPrefix: strings.ToUpper(App.Name) + "_",
+}
+
+// initiate dir structure on app initialization
+func init() {
+	err := WorkDir.Create()
+	if err != nil {
+		log.Fatal(err.Error())
+		os.Exit(1)
```
Comment (akshaypatidar1999) [inline app/structure.go:44]: Remove `os.Exit(1)` as it is unreachable. `log.Fatal` exists the app

File: app/structure.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+// WorkDir interface
+var WorkDir = workdir{
+	Location:     path.Join(os.Getenv("HOME"), "."+App.Name),
+	ConfigFile:   "config",
+	EnvVarPrefix: strings.ToUpper(App.Name) + "_",
+}
+
+// initiate dir structure on app initialization
+func init() {
```
Comment (akshaypatidar1999) [inline app/structure.go:40]: We do not need to do this on every app initialization. Lets do this only in the `configure` command

File: cmd/configure.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
+package cmd
```
Comment (akshaypatidar1999) [inline cmd/configure.go:1]: Keep this command in `cmd/configure/configure.go`

File: cmd/configure.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+	if odinSecretAccessKey != "" {
+		config.Keys.SecretAccessKey = odinSecretAccessKey
+	} else if os.Getenv("ODIN_SECRET_ACCESS_KEY") != "" {
+		config.Keys.SecretAccessKey = os.Getenv("ODIN_SECRET_ACCESS_KEY")
+	} else {
+		log.Fatal("Neither secret-access-key flag is passed nor environment variable ODIN_SECRET_ACCESS_KEY is set. Please either pass your access key using odin-secret-access-key or set your access key in ODIN_SECRET_ACCESS_KEY environment variable")
+	}
+
+	sha256 := sha256.New()
```
Comment (akshaypatidar1999) [inline cmd/configure.go:84]: Rename variable name as it conflicts with package name `sha256`

File: cmd/configure.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+func init() {
+	configureCmd.Flags().StringVar(&odinAccessKey, "access-key", "", "odin access key")
+	configureCmd.Flags().StringVar(&odinSecretAccessKey, "secret-access-key", "", "odin secret access key")
+	configureCmd.Flags().StringVar(&odinBackendAddress, "backend-address", "", "odin backend address with port")
+	configureCmd.Flags().BoolVarP(&odinInsecure, "insecure", "I", false, "odin insecure")
+	RootCmd.AddCommand(configureCmd)
+}
+
+func execute(cmd *cobra.Command) {
```
Comment (akshaypatidar1999) [inline cmd/configure.go:44]: This is a huge function, break it into smaller ones

File: cmd/configure.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+	configPath := path.Join(app.WorkDir.Location, app.WorkDir.ConfigFile)
+	if _, err := os.Stat(configPath); os.IsNotExist(err) {
```
Comment (akshaypatidar1999) [inline cmd/configure.go:46]: Creating a file if it does not exist is a utility method, move this to `pkg`.

File: cmd/configure.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		config.BackendAddr = odinBackendAddress
+	} else if os.Getenv("ODIN_BACKEND_ADDRESS") != "" {
+		config.BackendAddr = os.Getenv("ODIN_BACKEND_ADDRESS")
+	} else {
+		config.BackendAddr = DEFAULT_BACKEND_ADDR
+	}
+
+	config.Insecure = odinInsecure
+
+	if odinAccessKey != "" {
```
Comment (akshaypatidar1999) [inline cmd/configure.go:68]: This code block is very similar and is repeated 3 times, see if you can generalise this. Take flag name, env variable name and return value based on precedece

File: proto/dream11/od/auth/v1/auth.proto [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+option go_package = "github.com/dream11/odin/proto/gen/go/dream11/od/auth/v1";
+
+service AuthService {
+  rpc GetUserToken(GetUserTokenRequest) returns (GetUserTokenResponse) {}
+}
+
+message GetUserTokenRequest {
+  string client_id = 1;
+  string client_secret_hash = 2;
```
Comment (akshaypatidar1999) [inline proto/dream11/od/auth/v1/auth.proto:12]: Let's take plain secret also as plain text from user, odin cli will create a hash from it and send it to deployer

File: app/structure.go (hunk 2 above)
Comment (rahulrayal) [reply app/structure.go]: removed it

File: cmd/configure.go (hunk 3 above)
Comment (rahulrayal) [reply cmd/configure.go]: kept it

File: cmd/configure.go (hunk 4 above)
Comment (rahulrayal) [reply cmd/configure.go]: renamed it

File: cmd/configure.go (hunk 5 above)
Comment (rahulrayal) [reply cmd/configure.go]: broke it

File: cmd/configure.go (hunk 6 above)
Comment (rahulrayal) [reply cmd/configure.go]: moved it

File: cmd/configure.go (hunk 7 above)
Comment (rahulrayal) [reply cmd/configure.go]: generalised it

File: proto/dream11/od/auth/v1/auth.proto (hunk 8 above)
Comment (rahulrayal) [reply proto/dream11/od/auth/v1/auth.proto:12]: It is like this only. for the backend it is client_secret_hash (after hashing the secret-access-key set by the user)

File: cmd/configure/configure.go [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+	} else if configValue != "" {
+		return configValue
+	} else if defaultValue != "" {
+		return defaultValue
+	} 
+	log.Fatalf("Please pass %s flag nor set environment variable %s to configure", flagKey, envVariableName)
+	return ""
+}
+
+func setConfig(profile string, config *configuration.Configuration) {
```
Comment (akshaypatidar1999) [inline cmd/configure/configure.go:98]: Rename this to `writeConfig` and move to `pkg/config`

File: cmd/configure/configure.go [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+	config := viperConfig.GetConfig()
+	profile := viper.GetString("profile")
+
+	config.BackendAddress = getConfigKey("backend-address", odinBackendAddress, "ODIN_BACKEND_ADDRESS", config.BackendAddress, defaultBackendAddress)
+	config.Insecure = odinInsecure
+	config.Keys.AccessKey = getConfigKey("access-key", odinAccessKey, "ODIN_ACCESS_KEY", config.Keys.AccessKey, "")
+	config.Keys.SecretAccessKey = getConfigKey("secret-access-key", odinSecretAccessKey, "ODIN_SECRET_ACCESS_KEY", config.Keys.SecretAccessKey, "")
+	setConfig(profile, config)
+
+	hash := sha256.New()
```
Comment (akshaypatidar1999) [inline cmd/configure/configure.go:67]: Move this to a method

File: cmd/configure/configure.go [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+func init() {
+	configureCmd.Flags().StringVar(&odinAccessKey, "access-key", "", "odin access key")
+	configureCmd.Flags().StringVar(&odinSecretAccessKey, "secret-access-key", "", "odin secret access key")
+	configureCmd.Flags().StringVar(&odinBackendAddress, "backend-address", "", "odin backend address with port")
+	configureCmd.Flags().BoolVarP(&odinInsecure, "insecure", "I", false, "odin insecure")
+	cmd.RootCmd.AddCommand(configureCmd)
+}
+
+func execute(cmd *cobra.Command) {
+	dirPath:=path.Join(os.Getenv("HOME"), "."+app.App.Name)
```
Comment (akshaypatidar1999) [inline cmd/configure/configure.go:49]: Proper spacing

File: cmd/configure/configure.go [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	config := viperConfig.GetConfig()
+	profile := viper.GetString("profile")
+
+	config.BackendAddress = getConfigKey("backend-address", odinBackendAddress, "ODIN_BACKEND_ADDRESS", config.BackendAddress, defaultBackendAddress)
+	config.Insecure = odinInsecure
+	config.Keys.AccessKey = getConfigKey("access-key", odinAccessKey, "ODIN_ACCESS_KEY", config.Keys.AccessKey, "")
+	config.Keys.SecretAccessKey = getConfigKey("secret-access-key", odinSecretAccessKey, "ODIN_SECRET_ACCESS_KEY", config.Keys.SecretAccessKey, "")
+	setConfig(profile, config)
```
Comment (akshaypatidar1999) [inline cmd/configure/configure.go:65]: Remove this as the config is written later

File: cmd/configure/configure.go [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+	if err := dir.CreateDirIfNotExist(dirPath); err != nil {
+		log.Fatalf("Error creating the .%s folder: %v", app.App.Name, err)
+	}
+	configPath := path.Join(dirPath, "config")
+	if err := dir.CreateFileIfNotExist(configPath); err != nil {
+		log.Fatal("Error creating the config file")
+	}
+
+	config := viperConfig.GetConfig()
+	profile := viper.GetString("profile")
```
Comment (akshaypatidar1999) [inline cmd/configure/configure.go:59]: Get profile also inside `writeConfig` method

File: odin
Comment (boppanasusanth) [inline odin:1]: remove this file.

File: cmd/configure/configure.go (hunk 10 above)
Comment (rahulrayal) [reply cmd/configure/configure.go]: Done

File: cmd/configure/configure.go (hunk 11 above)
Comment (rahulrayal) [reply cmd/configure/configure.go]: Done

File: cmd/configure/configure.go (hunk 12 above)
Comment (rahulrayal) [reply cmd/configure/configure.go]: Done

File: odin
Comment (rahulrayal) [reply odin:1]: Done

**Pull request merged**